ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/BlockTree.hs

```diff
@@ -104,14 +103,25 @@ findFragment :: AF.HasHeader blk => AF.Point blk -> BlockTree blk -> Maybe (AF.A
 findFragment point blockTree =
   fst <$> firstJust (map (\fragment -> AF.splitAfterPoint fragment point) $ allFragments blockTree)
 
-newtype InvolvesRollback = InvolvesRollback Bool
-
-findPath :: AF.HasHeader blk => AF.Point blk -> AF.Point blk -> BlockTree blk -> Maybe (InvolvesRollback, AF.AnchoredFragment blk)
+-- | @findPath source target blockTree@ finds a path from the @source@ point to
+-- the @target@ point in the @blockTree@, or return @Nothing@. A path is a
+-- fragment anchored at the @source@ or an ancestor of the @source@. One can
+-- distinguish three cases:
+--
+-- - the fragment is anchored at the @source@: all the blocks are descendants of
+--   the source; serving this fragment would only require rolling forward;
+--
+-- - the fragment is not anchored at the @source@ and is empty: it is then in
+--   fact anchored at the @target@ which is an ancestor of the source;
+--
+-- - the fragment is not anchored at the @source@ and is not empty: serving this
+--   fragment would require rolling backwards to the anchor.
+findPath :: AF.HasHeader blk => AF.Point blk -> AF.Point blk -> BlockTree blk -> Maybe (AF.AnchoredFragment blk)
```

Comment:
Maybe the return type could be enriched to make the three cases explicit:
```haskell
data FindPath blk
  = PathAnchoredAtSource (AF.AnchoredFragment blk)
  | PathTargetIsAncestorOfSource
  | PathWithSourceRollback (AF.AnchoredFragment blk)
```

Reply:
I love this! It turned out that this was not so usable after all; it would have been with a way to return non-empty fragments explicitly. Still, I think it is good to return at least the information about whether the path is anchored at the source or not. I have documented much better (I hope) and introduced this new return type in 47331e318.